alc/backends/coreaudio.cpp

```diff
@@ -271,6 +271,47 @@ void EnumerateDevices(std::vector<DeviceEntry> &list, bool isCapture)
     newdevs.swap(list);
 }
 
+struct DeviceHelper
+{
+public:
+    DeviceHelper()
+    {
+        AudioObjectPropertyAddress addr = {kAudioHardwarePropertyDefaultOutputDevice,
+                                           kAudioObjectPropertyScopeGlobal, kAudioObjectPropertyElementMain};
+        OSStatus status = AudioObjectAddPropertyListener(kAudioObjectSystemObject, &addr, DeviceListenerProc, nil);
+        if (status != noErr)
+            ERR("AudioObjectAddPropertyListener fail: %d", status);
+    }
+    ~DeviceHelper()
+    {
+        AudioObjectPropertyAddress addr = {kAudioHardwarePropertyDefaultOutputDevice,
+                                           kAudioObjectPropertyScopeGlobal, kAudioObjectPropertyElementMain};
+        OSStatus status = AudioObjectRemovePropertyListener(kAudioObjectSystemObject, &addr, DeviceListenerProc, nil);
+        if (status != noErr)
+            ERR("AudioObjectRemovePropertyListener fail: %d", status);
+    }
+
+    static OSStatus DeviceListenerProc(AudioObjectID inObjectID,
+                                        UInt32 inNumberAddresses,
+                                        const AudioObjectPropertyAddress *inAddresses,
+                                        void *__nullable inClientData)
+    {
+        for (UInt32 i = 0; i < inNumberAddresses; ++i)
+        {
+            switch (inAddresses[i].mSelector)
+            {
+            case kAudioHardwarePropertyDefaultOutputDevice:
+            case kAudioHardwarePropertyDefaultSystemOutputDevice:
+            case kAudioHardwarePropertyDefaultInputDevice:
+                alc::Event(alc::EventType::DefaultDeviceChanged, "Default device changed: "+std::to_string(inAddresses[i].mSelector));
+                break;
+            }
+        }
+    }
+};
+
+static std::unique_ptr<DeviceHelper> sDeviceHelper;
```

Comment:
Might make more sense to use `std::optional`, to dynamically initialize it without needing dynamic allocation for an empty struct.

Reply:
done